### DISCUSSION TITLE
Testing Pipe and Injector

### DISCUSSION THREAD

**Discussion by josejachuf** [emmett-framework/emmett#274] [Q&A]
Hi @gi0baro I ask around here because apparently weppy-talk is dead.

This is in weppy 1.3.4 and python 3.6

Continuing with the tests, I am trying to test a Pipe and an Injector. I am sure that both work well, because I have been using them for years and they have not touched since that time. In this case it is simply to understand how the test would be.

```
class SignInjector(Injector):
    @staticmethod
    def sign(path, args, params, anchor, scheme, host, language):

        s = Signer(KEY)
        builder = RouteUrl([url(path)])
        _url = builder.url(None, None, language, args, params, None)

        p = _url.find('_signature')
        if p > 0:
            _url = _url[:p-1] + _url[p+38:]

        _url = str.encode(_url)
        signature = s.get_signature(_url).decode("utf-8")

        return signature


class SignPipe(Pipe):

    def pipe(self, next_pipe, **kwargs):
        if self.valid_url():
            return next_pipe(**kwargs)
        abort(404)

    def valid_url(self):
        signature = request.query_params._signature or ''
        builder = RouteUrl([request.path_info])
        _url = builder.url(None, None,
                           request.language, [],
                           {key: request.query_params[key] for key in request.query_params.keys() if key != '_signature'},
                           None)

        s = Signer(KEY)
        sig = str.encode(signature)
        value = str.encode(_url)
        return s.verify_signature(value=value, sig=sig)

```

I think I could do it directly against the real app, but want to abstract me from it.
I am trying to create some tests seeing in [1] and [2]

I have the following (what is a copy of your test):

```
import pytest

from weppy import App, request, url
from weppy.globals import current
from weppy.testing.env import EnvironBuilder

from aplicacion.modules.signature import SignInjector, SignPipe

@pytest.fixture(scope='module')
def app():
    app = App(__name__)

    return app


def test_signed_route(app):
    builder = EnvironBuilder('/test_route/a/b')
    current.initialize(builder.get_environ())

    @app.route('/test_route/<str:a>/<str:b>')
    def test_route(a, b):
        return "Test Router"

    response = current.response
    app.route.dispatch()
    assert response.status != 404
    assert response.output == 'Test Router'

```

The above obviously passes the test, but what I want to have is something like this:

```
def test_signed_route(app):
    _url = url('test_route', args=['a', 'b'], params=dict(id=10), sign=SignInjector.sign)
    builder = EnvironBuilder(_url)
    current.initialize(builder.get_environ())

    @app.route('/test_route/<str:a>/<str:b>', pipeline=[SignPipe()])
    def test_route(a, b):
        return "Test Router"

    response = current.response
    app.route.dispatch()
    assert response.status != 404
    assert response.output == 'Test Router'
```

> raise RuntimeError('invalid url("%s",...)' % path)
> E               RuntimeError: invalid url("myapp.tests.test_signed_url.test_route",...)

So how should I approach this test?

[1] https://github.com/emmett-framework/emmett/blob/1.x/tests/test_pipeline.py
[2] https://github.com/emmett-framework/emmett/blob/1.x/tests/test_expose.pytest

**Comment by gi0baro** [comment]
@josejachuf probably you just need to call `url` after defining the route, eg:

```python
def test_signed_route(app):
    @app.route('/test_route/<str:a>/<str:b>', pipeline=[SignPipe()])
    def test_route(a, b):
        return "Test Router"

    _url = url('test_route', args=['a', 'b'], params=dict(id=10), sign=SignInjector.sign)
    builder = EnvironBuilder(_url)
    current.initialize(builder.get_environ())
    response = current.response
    app.route.dispatch()
    assert response.status != 404
    assert response.output == 'Test Router'
```

**Comment by josejachuf** [comment]
@gi0baro 

It does not work. Keep giving the same error.
If I change: 
`_url = url('test_route', args=['a', 'b'], params=dict(id=10), sign=SignInjector.sign)`
to:
`_url = url('test_route/a/b', params=dict(id=10), sign=SignInjector.sign)`

I get a different error:
> cls = <class 'weppy.expose.Expose'>
> 
>     @classmethod
>     def dispatch(cls):
>         #: get the right exposed function
>         request = current.request
>         route, reqargs = cls.match(request)
>         if not route:
> >           raise HTTP(404, body="Resource not found\n")
> E           weppy.http.HTTP: 404
> 
> ../../../venv36/lib/python3.6/site-packages/weppy/expose.py:300: HTTP
